test/torch_optimizer_test.py

```diff
@@ -0,0 +1,525 @@
+        return SimpleDataset(x, y)
+
+# Prepare the problem  to be solved
+def problem_setup():
+    bf.init()
+    num_epochs = 50
+    batch_size = 128
+    num_train_per_node = 1024
+    num_test_per_node = 128
+    lr = 0.05
+
+    # Setup Problem
+    problem_builder = LinearProblemBuilder()
+    train_dataset = problem_builder.get_dataset(num_train_per_node)
+    train_dataloader = DataLoader(train_dataset, batch_size=batch_size)
+    test_dataset = problem_builder.get_dataset(num_test_per_node)
+    test_dataloader = DataLoader(test_dataset, batch_size=batch_size)
+    # Setup Model
+    model = LinearNet(problem_builder.input_dim, problem_builder.output_dim)
+    # Setup Optimizer
+    optimizer = optim.Adam(model.parameters(), lr=lr*bf.size())
+    bf.broadcast_parameters(model.state_dict(), root_rank=0)
+    bf.broadcast_optimizer_state(optimizer, root_rank=0)
+    return problem_builder, train_dataloader, test_dataloader, model, optimizer, num_epochs
+
+# Standard training process
+def standard_train(model, optimizer, dataloader, isCUDA):
+    mseloss = nn.MSELoss()
+    model.train()
+    for data, target in dataloader:
+        if isCUDA:
+            data, target = data.cuda(), target.cuda()
+        y = model(data)
+        loss = mseloss(y, target)
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+# Dynamic topology training process for win_put
+def dynamic_win_put_train(model, optimizer, dataloader, isCUDA, epoch):
+    mseloss = nn.MSELoss()
+    model.train()
+    for batch_idx, (data, target) in enumerate(dataloader):
+        if epoch < 3:
+            return
+        num_out_neighbors = len(bf.out_neighbor_ranks())
+        sent_neighbor = bf.out_neighbor_ranks()[batch_idx % num_out_neighbors]
+        optimizer.dst_weights = {sent_neighbor: 1.0}
+        if isCUDA:
+            data, target = data.cuda(), target.cuda()
+        y = model(data)
+        loss = mseloss(y, target)
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+# Dynamic topology training process for neighbor_allreduce
+def dynamic_neighbor_allreduce_train(model, optimizer, dataloader, isCUDA, dynamic_topo_gen):
+    mseloss = nn.MSELoss()
+    model.train()
+    for data, target in dataloader:
+        send_neighbors, recv_neighbors = next(dynamic_topo_gen)
+        optimizer.send_neighbors = send_neighbors
+        optimizer.neighbor_weights = {r: 1/(len(recv_neighbors) + 1) for r in recv_neighbors}
+        optimizer.self_weight = 1 / (len(recv_neighbors) + 1)
+
+        if isCUDA:
+            data, target = data.cuda(), target.cuda()
+        y = model(data)
+        loss = mseloss(y, target)
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+# Training process with mini_batch, expecting optimizer skipping communication
+def skip_communication_train(model, optimizer, dataloader, isCUDA, mini_batch_size):
+    mseloss = nn.MSELoss()
+    model.train()
```

Comment:
Add a todo here or Github issue about the potential problem that if the batch size cannot be divided by J so that some left-over data may never be communicated?

Reply:
Done.